llvm/lib/YkIR/YkIRWriter.cpp

```diff
@@ -272,6 +293,13 @@ class YkIRWriter {
   vector<llvm::Constant *> Constants;
   vector<llvm::GlobalVariable *> Globals;
 
+  // Line-level debug line info for the instructions of the module.
+  //
+  // If debug info is not being compiled in, this will be empty.
+  map<InstID, LineInfo> LineInfos;
```

Comment:
I might add something like `OPT: This should/could be a vector` here to remind us that `map` isn't the ideal long-term format here.

Reply:
This is tricky and I had to think about this a bit. Many consecutive AOT IR instructions all map to the same source line. To prevent repeatedly printing the same source line when dumping the AOT module, I opted to only record the source line for the first AOT instruction in a run. Once you've done that, the mapping becomes sparse, and a vector is then wasteful, as you'd have to pad it with `None`s.

Dunno if that makes sense?

Reply:
Yes, this is a well known problem (or... at least I know about it!). You're right that a sparse `Vec` is wasteful *but* it's probably less wasteful than a map! It also allows you to later compress the backing. For example, Converge uses (in essence) a simple run-length encoding to turn each repeated entry into a single bit (well, not quite that good, but that's the general idea).

However, provided you're comfortable that we can change the `map` to something else without changing the way we expose things, I think we can stick with this for now.

Reply:
I don't think it'll be too hard to change later. Like I said, I've not focused on optimisation much here yet.

Reply:
OK.

Reply:
Note that this is an AOT compile time cost too.

Reply:
Ah, we don't deserialise it in yk as a `HashMap`?

Reply:
We are just storing the info we need to emit the debug information last. At that point we can choose to serialise this however we want.

I don't think deku has a defined way to deserialise hashmaps. I've just encoded it like
```
num_records
key0
val0
...
keyn
valn
```

in the runtime, I had deku load those into a flat vector, which we then map to (currently) a hashmap. It can all be changed later as we see fit, I just wanted something to iterate upon.

Reply:
OK.